File: src/lowlevel/client/icefsoperators/icefsRename.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:41:01
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsRename.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsRename(fuse_req_t fuseReq, fuse_ino_t parentInode,
                                const char *name, fuse_ino_t newParent,
                                const char *newName, unsigned int flags) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsRenameReq req;
      icefsgrpc::IcefsRenameRes res;
      grpc::ClientContext ctx;

      req.set_parent_inode(parentInode);
      req.set_name(name);
      req.set_new_parent_inode(newParent);
      req.set_new_name(newName);
      req.set_flags(flags);

      grpc::Status status = gRpcClient->DoIcefsRename(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_err(fuseReq, ICEFS_EOK);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsRenameReq req;
      icefsthrift::IcefsRenameRes res;

      req.__set_parent_inode(parentInode);
      req.__set_name(name);
      req.__set_new_parent_inode(newParent);
      req.__set_new_name(newName);
      req.__set_flags(flags);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsRename(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      fuse_reply_err(fuseReq, res.status);
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}

File: src/lowlevel/server/icefsoperators/icefsRename.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 13:44:09
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsRename.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsRename(parentFakeInode uint64, name string, newParentFakeInode uint64, newName string, flags uint32) (status int32) {
	var inode *IcefsInode
	var newInode *IcefsInode
	var err error

	if flags != 0 {
		status = int32(syscall.EINVAL)
		return
	}

	s.inodeCacheLock.RLock()
	inode = s.getIcefsInode(parentFakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}

	newInode = s.getIcefsInode(newParentFakeInode)
	if newInode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}

	inode.inodeLock.RLock()
	newInode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	err = syscall.Renameat(inode.fd, name, newInode.fd, newName)
	inode.inodeLock.RUnlock()
	newInode.inodeLock.RUnlock()

	status = icefserror.IcefsStdErrno(err)
	return
}

func (s *IcefsGRpcServer) DoIcefsRename(ctx context.Context, req *pb.IcefsRenameReq) (*pb.IcefsRenameRes, error) {
	var res pb.IcefsRenameRes

	res.Status = s.server.doIcefsRename(req.ParentInode, req.Name, req.NewParentInode, req.NewName, req.Flags)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsRename(ctx context.Context, req *icefsthrift.IcefsRenameReq) (*icefsthrift.IcefsRenameRes, error) {
	var res icefsthrift.IcefsRenameRes

	res.Status = s.server.doIcefsRename(uint64(req.ParentInode), req.Name, uint64(req.NewParentInode_), req.NewName_, uint32(req.Flags))

	return &res, nil
}
